### PR TITLE
update publicips service to use Azure SDK v2

### LINKED ISSUE #3919: Update publicips service to SDK v2

### PR DESCRIPTION
**Pull request by nawazkh**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
- This PR will update publicips service to use Azure SDK v2

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3919 

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
update publicips service to use Azure SDK v2
```

### CODE REVIEW

**Comment by nawazkh** [comment]
Current PR is dependent on https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3969 so I will be cherry-picking the changes from https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3969 onto current PR to resolve assignment error at https://github.com/nawazkh/cluster-api-provider-azure/blob/c20099b2b691143f1034ef8b9309ed932c8e1b7e/azure/services/loadbalancers/spec.go#L146

File: azure/services/publicips/spec.go
Comment (nawazkh) [inline azure/services/publicips/spec.go:100]: `// s.FailureDomains needs to be updated to a []*string`
This is more of an informational comment while removing. Will remove after review.

**Comment by nawazkh** [comment]
/hold
Merge https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3969 before merging this PR

File: azure/converters/extendedlocation.go
Comment (nawazkh) [inline azure/converters/extendedlocation.go:38]: We no longer need this function, will delete it.

Comment (nawazkh) [inline azure/converters/extendedlocation.go:38]: I dont really need to touch this function. Undoing the change.

File: azure/converters/extendedlocation_test.go
Comment (nawazkh) [inline azure/converters/extendedlocation_test.go:65]: I don't need to update this test either in this PR. Will follow up in another PR.

File: azure/services/publicips/spec.go
Comment (nawazkh) [inline azure/services/publicips/spec.go:100]: This is the reason for touching so many mocks.

Comment (nawazkh) [reply azure/services/publicips/spec.go:100]: I can instead add a converter here that converts `[]string` to `[]*string` and leave all the mocks as is..

File: azure/services/publicips/client.go
Comment (nojnhuh) [inline azure/services/publicips/client.go:34]: It looks like `auth` isn't used.

File: azure/services/publicips/client.go
Comment (nojnhuh) [inline azure/services/publicips/client.go:72]:
```suggestion
		return nil, nil, errors.Errorf("%T is not an armnetwork.PublicIPAddress", parameters)
```

File: azure/services/publicips/spec.go
Comment (nojnhuh) [inline azure/services/publicips/spec.go:63]:
```suggestion
			return nil, errors.Errorf("%T is not an armnetwork.PublicIPAddress", existing)
```

File: azure/services/publicips/spec.go
Comment (nojnhuh) [inline azure/services/publicips/spec.go:93]: It looks like this was the last remaining instance of the "v1" version of this function, so we can probably remove that now and rename the v2 one to drop the "v2"

File: azure/services/publicips/client.go
Comment (nawazkh) [reply azure/services/publicips/client.go]: Updated 7ba1619f6df44b7cc1ffb572c216a18297a7c4ed.

File: azure/services/publicips/client.go
Comment (nawazkh) [reply azure/services/publicips/client.go]: Updated via 7ba1619f6df44b7cc1ffb572c216a18297a7c4ed

File: azure/services/publicips/spec.go
Comment (nawazkh) [reply azure/services/publicips/spec.go]: Updated via 7ba1619f6df44b7cc1ffb572c216a18297a7c4ed

File: azure/services/publicips/spec.go
Comment (nawazkh) [reply azure/services/publicips/spec.go]: Updated via 7ba1619f6df44b7cc1ffb572c216a18297a7c4ed

**Review by nojnhuh — COMMENTED** [review]
/lgtm

File: azure/services/publicips/client.go
Comment (mboersma) [inline azure/services/publicips/client.go:45]:
```suggestion
		return nil, errors.Wrap(err, "failed to create armnetwork client factory")
```

**Review by mboersma — COMMENTED** [review]
This looks great! I only found two very nitpicky string changes for consistency with other services.

File: azure/services/publicips/client.go
Comment (nawazkh) [reply azure/services/publicips/client.go]: Updated via bc9a6ceb308c29b36642dc2479e31de195a761d9

File: azure/services/publicips/client.go
Comment (nawazkh) [reply azure/services/publicips/client.go]: Updated via bc9a6ceb308c29b36642dc2479e31de195a761d9

**Review by mboersma — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
